src/history/hs_rec.c

```diff
@@ -213,8 +229,23 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
          */
         if (hs_btree_id == btree->id && start_time_point->ts == hs_start_ts) {
             WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+
+#ifdef HAVE_DIAGNOSTIC
+            if (cmp == 0) {
+                WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+                  &upd_type_full_diag, existing_val));
+                WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
+                if (cmp == 0)
+                    WT_ASSERT(session,
+                      start_time_point->txnid == WT_TXN_NONE ||
+                        start_time_point->txnid != hs_cbt->upd_value->tw.start_txn ||
+                        start_time_point->ts != hs_cbt->upd_value->tw.start_ts);
+                counter = hs_counter + 1;
+            }
```

Comment:
I think it is worth running a full set of stress tests and mongodb debug tests to make sure this doesn't crash on any corner cases.

Reply:
Agree. I'll start patch builds and add links to the ticket.

Reply:
@quchenhao @bvpvamsikrishna - The MongoDB patch looks good.